test/test_autoextract.py

```diff
@@ -0,0 +1,108 @@
+import pytest
+
+from w3lib.http import basic_auth_header
+from scrapy.http import Request, Response
+from scrapy.spiders import Spider
+from scrapy.utils.test import get_crawler
+
+from scrapy_autoextract import AutoExtractMiddleware
+from scrapy_autoextract.middlewares import AutoExtractConfigError
+
+AUTOX_META = {'autoextract': {'enabled': True}}
+
+MW_SETTINGS = {
+    'AUTOEXTRACT_USER': 'apikey',
+    'AUTOEXTRACT_PAGE_TYPE': 'article',
+}
+
+
+def setup_module(module):
+    global spider
+    spider = Spider('spidr')
+
+
+def _mock_mw(spider, settings=None):
+
+    class MockedDownloader:
+        slots = {}
+
+        def _get_slot_key(self, a, b):
+            return str(a) + str(b)
+
+    class MockedEngine:
+        downloader = MockedDownloader()
+        fake_spider_closed_result = None
+
+        def close_spider(self, spider, reason):
+            self.fake_spider_closed_result = (spider, reason)
+
+    # with `spider` instead of `type(spider)` raises an exception
+    crawler = get_crawler(type(spider), settings)
+    crawler.engine = MockedEngine()
+    return AutoExtractMiddleware.from_crawler(crawler)
+
+
+def _assert_disabled(spider, settings=None):
+    mw = _mock_mw(spider, settings)
+    req = Request('http://quotes.toscrape.com', meta=AUTOX_META)
+    out = mw.process_request(req, spider)
+    assert out is None
+    assert req.meta.get('autoextract') is None
+    res = Response(req.url, request=req)
+    assert mw.process_response(req, res, spider) == res
+
+
+def _assert_enabled(spider,
+                    settings=None,
+                    url='http://quotes.toscrape.com',
+                    api_url='autoextract.scrapinghub.com',
+                    api_auth=basic_auth_header('apikey', '')):
+    mw = _mock_mw(spider, settings)
+
+    req = Request(url, meta=AUTOX_META)
+    out = mw.process_request(req, spider)
+    assert api_url in out.url
+    assert out.meta['autoextract'].get('enabled')
+    assert out.headers.get('Authorization') == api_auth
+
+    resp = Response(out.url, request=out)
+    resp._body = '[{}]'  # workaround to make the body work with Python 3.5+
```

Comment:
It doesn't look good; I think tests showed you a real issue in Python 3.5. In Scrapy response.body is bytes, response.text is text. In case of JSON you probably need to call `response.body.decode('utf8')`, and pass it to json.loads.